### ISSUE TITLE
2.16.0 - Issue with persisting entities

### ISSUE THREAD

**Issue by dmaicher** [doctrine/doctrine2#10869]
### BC Break Report

|    Q        |   A
|------------ | ------
| BC Break    | yes
| Version     | 2.16.0

#### Summary

I have a case in our test suite where entities are not persisted properly anymore. This worked fine on 2.15.

The insert statement does not bind any parameters and fails:

```
PDOException: SQLSTATE[42000]: Syntax error or access violation: 1064 You have an error in your SQL syntax; check the manual that corresponds to your MySQL server version for the right syntax to use near '?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)' at line 1

/var/www/app/symfony/vendor/doctrine/dbal/src/Driver/PDO/Statement.php:121
/var/www/app/symfony/vendor/doctrine/dbal/src/Driver/Middleware/AbstractStatementMiddleware.php:69
/var/www/app/symfony/vendor/doctrine/dbal/src/Logging/Statement.php:98
/var/www/app/symfony/vendor/doctrine/dbal/src/Driver/Middleware/AbstractStatementMiddleware.php:69
/var/www/app/symfony/vendor/symfony/doctrine-bridge/Middleware/Debug/Statement.php:68
/var/www/app/symfony/vendor/doctrine/dbal/src/Statement.php:190
/var/www/app/symfony/vendor/doctrine/dbal/src/Statement.php:249
/var/www/app/symfony/vendor/doctrine/orm/lib/Doctrine/ORM/Persisters/Entity/BasicEntityPersister.php:280
/var/www/app/symfony/vendor/doctrine/orm/lib/Doctrine/ORM/UnitOfWork.php:1177
/var/www/app/symfony/vendor/doctrine/orm/lib/Doctrine/ORM/UnitOfWork.php:442
/var/www/app/symfony/vendor/doctrine/orm/lib/Doctrine/ORM/EntityManager.php:403
``` 

The logs (see `parameters: []`):

```
10:11:57 DEBUG     [doctrine] Executing statement: INSERT INTO widget_settings (hash, enabled, language, orientation, position, widget_content, tooltip_content, header_color, detail_urls, left_title, use_left_title, right_title, use_right_title, logo, use_logo, is_primary, booking_btn_color, widget_name, booking_urls, review_config_id, account_id, parent_id) VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?) (parameters: [], types: [])
``` 

#### How to reproduce

I cannot provide a small reproducer yet. I had a look and somehow inside the `BasicEntityPersister` the call to `prepareInsertData($entity)` with that entity returns an empty array.

The setup part of the test code that triggers the error looks like this:

```php
        $account = $this->fixtures->createAccount();
        $this->em->persist($account);
        $this->em->persist($this->fixtures->createWidgetSettings($account, false));
        $this->em->persist($this->fixtures->createWidgetSettings($account, false));

        $this->em->flush();
``` 

The issue seems to come from the second `$this->em->persist($this->fixtures->createWidgetSettings($account, false));` line :thinking: 

@mpdude does this information help at all to pinpoint the issue? I can try to provide a small reproducer

**Comment by dmaicher** [comment]
I did some more digging and this seems to be caused by some event listener that does stuff on `postPersist` and then calls `$em->flush()` itself.

**Comment by derrabus** [comment]
Can you create a reproducer for this particular issue?

**Comment by mpdude** [comment]
Similar report here: https://github.com/doctrine/orm/pull/10547#issuecomment-1657188129

Not sure if it is allowed/supported to call `flush()` from a `postPersist` listener?

**Comment by dmaicher** [comment]
> Not sure if it is allowed/supported to call `flush()` from a `postPersist` listener?

yeah thats the question :smile: Nothing is mentioned in the [docs](https://www.doctrine-project.org/projects/doctrine-orm/en/2.15/reference/events.html#postupdate-postremove-postpersist)

**Comment by dmaicher** [comment]
Well actually it says

> Changes in here **are not relevant to the persistence in the database**, but you can use these events to alter non-persistable items, like non-mapped fields, logging or even associated classes that are not directly mapped by Doctrine.

**Comment by derrabus** [comment]
Since `postPersist` is triggered _during_ a flush operation, it's at least a very bad idea to call `flush()` from there. 🙈

**Comment by mpdude** [comment]
https://github.com/doctrine/orm/commit/b42cf99402f5cd6a899fcbc69cc9072a084ceea5 migth be related, it makes a change to `UnitOfWork::executeInserts()`:

* Pre-#10547, it performs `INSERT`s for all entities of a particular class; after that, call `postPersist` handlers for all those entities. Then proceed to `INSERT`s for the next class.
* Post-#10547, `INSERT`s are no longer grouped by entity class (we may even have to go back and forth between different classes). Thus, `postPersist` is called immediately after every single entity that has been `INSERT`ed.

**Comment by dmaicher** [comment]
I will rewrite this listener logic on our end. So feel free to close this issue. I think its very tricky to support all those weird edge cases :stuck_out_tongue:

**Comment by mpdude** [comment]
Should we prevent `flush()` reentrance with an exception? Seems its behaviour has been changed during the course of #10547, but I am not sure if it was ever supposed to work/be supported in the first place. (@beberlei WDYT?)

I am really afraid what might happen when the UoW has computed its lists of insertions, field updates etc. and then, during the course of the flush, some event listener calls `flush()` again, causing the same structures to be updated. Fairly sure this never had defined semantics and/or test coverage.

**Comment by dmaicher** [comment]
So I had a look at rewriting the listener logic and its not easy :cry: I would need to hook into the `flush` process and persist + flush some additional entities but **only after** all previous flushes are done and entities have IDs.

Currently I don't see any way of achieving this then if I cannot `flush` with any of the events. Even `postFlush` seems to not really allow it as there is some cleanup stuff happening afterwards.

**Comment by mpdude** [comment]
Assuming the `postFlush` could be deferred to after the final cleanups... would it work then?

**Comment by dmaicher** [comment]
> Assuming the `postFlush` could be deferred to after the final cleanups... would it work then?

Indeed I just tested it: that would work :blush:

**Comment by mpdude** [comment]
@beberlei You documented in https://github.com/doctrine/orm/commit/b6c3fc5b1ab8f97ba3a47b5a667ef8986c48059e that `flush()` cannot be safely called from `postFlush` listeners.

Do you remember any particular reason for that apart from potential recursion issues?

**Comment by tasselchof** [comment]
@mpdude I will try to check it, but as I see it from my mobile phone your reproducer is for single key reference, I am using composite keys. Product offer is partitioned by $shop and $id. This might be the difference we are looking for.

**Comment by mpdude** [comment]
@tasselchof maybe you meant to comment on #10868?

**Comment by tasselchof** [comment]
> @tasselchof maybe you meant to comment on #10868?

Yes, sorry once more, didn’t saw it on mobile.

**Comment by alexander-schranz** [comment]
I tried to create a simple reproducer but currently failing with it, the thing I did not yet tried is when the entity is a `ResolvedTargetListener` related.

Atleast the flush on postPersist didn't have any issues here:

https://github.com/alexander-schranz/reproducer-doctrine-orm-issue-10869/blob/3ec32bb9c2a2b149a29027a302d071e45f54b6c4/src/EventListener/AccountNumberEventListener.php#L32

This test currently runs fine:

https://github.com/alexander-schranz/reproducer-doctrine-orm-issue-10869/blob/3ec32bb9c2a2b149a29027a302d071e45f54b6c4/tests/Repository/AccountRepositoryTest.php#L28

So maybe really more related to ResolvedTargetListener but would need to test that, if somebody got time before me try to reproduce it on a simpler installation maybe can try it to create on top.

**Comment by jmperruchini** [comment]
hello everyone, I have the same problem since the update of symfony 6.3.3..It worked very well before, I used postPersist to retrieve the generated ID and trace by creating a log with a flush ..

public function postPersist(PostPersistEventArgs $args): void
    {
        if($args->getObject() instanceof Log){
            return;
        }
        $log = new Log();
        $log->setActionLog('create');
        $log->setModuleLog('personnel');
        $log->setEntiteIdLog($args->getObject()->getId()));
        $args->getObjectManager()->persist($log);
        $args->getObjectManager()->flush();
    }

with the following result (incomplete)=>

An exception occurred while executing a query: SQLSTATE[42000]: Syntax error or access violation: 1064 You have an error in your SQL syntax; check the manual that corresponds to your MySQL server version for the right syntax to use near '?, ?, ?, ?, ?, ?, ?, ?, ?)' at line 1
Doctrine\DBAL\Exception\ SyntaxErrorException
Show exception properties

Doctrine\DBAL\Exception\SyntaxErrorException {#5000 ▼
  -query: Doctrine\DBAL\Query {#4993 ▼
    -sql: "INSERT INTO log (module_log, action_log, entite_type_log, entite_id_log, utilisateur_ip_log, libelle_log, created_at, utilisateur_id_log, patient_id_log) VALUES ▶"
    -params: []
    -types: []
  }
}

the error only exists for entities with associations !!

I tried to change the logic using onflush and postflush, but I can't get the newly created entity iD.
I am blocked !!

**Comment by mpdude** [comment]
@dmaicher this code

https://github.com/alexander-schranz/reproducer-doctrine-orm-issue-10869/blob/3ec32bb9c2a2b149a29027a302d071e45f54b6c4/src/EventListener/AccountNumberEventListener.php#L32

linked in [this comment](https://github.com/doctrine/orm/issues/10869#issuecomment-1666004010) uses a `postPersist` listener to make changes to the entity that has just been persisted, and then tries to `flush()` again.

Is your use case the same? Or are you changing/adding other objects than the one the `postPersist` has been dispatched for?

**Comment by mpdude** [comment]
@alexander-schranz I wanted to suggest you could try using `onflush`, since that can make changes to entity values just in time (https://www.doctrine-project.org/projects/doctrine-orm/en/2.16/reference/events.html#onflush).

But then I realized you need the database-provided ID to fill in the missing object value, so you have to `INSERT` first before the update can be made.

**Comment by mpdude** [comment]
@alexander-schranz Your code seems to be working in 2.16.x, see #10906. So, there must be more to it...?

**Comment by dmaicher** [comment]
> @dmaicher this code
> 
> https://github.com/alexander-schranz/reproducer-doctrine-orm-issue-10869/blob/3ec32bb9c2a2b149a29027a302d071e45f54b6c4/src/EventListener/AccountNumberEventListener.php#L32
> 
> linked in [this comment](https://github.com/doctrine/orm/issues/10869#issuecomment-1666004010) uses a `postPersist` listener to make changes to the entity that has just been persisted, and then tries to `flush()` again.
> 
> Is your use case the same? Or are you changing/adding other objects than the one the `postPersist` has been dispatched for?

@mpdude my case is different: I persist new entities (of the same class) that have a relation to the entity for which `postPersist` was triggered.

**Comment by antoinemineau** [comment]
I confirm having the same issue.

We have flush in PostPersist which is probably not recommended but was working fine in 2.15.

In 2.16, We have INSERT with no parameters
In our listeners we create new entities or update the persisted one, both breaks by creating insert with empty parameters.

![Screenshot 2023-08-11 at 19 41 36](https://github.com/doctrine/orm/assets/520893/cf06eafe-79e4-498c-89a8-494577389b6b)

**Comment by tristanbes** [comment]
Experiencing the same through DoctrineExtensions, timestampable is no longer applied with doctrine/orm 2.16 causing DB exceptions

**Comment by dbannik** [comment]
I feel the same after the update
![image](https://github.com/doctrine/orm/assets/16108701/500350be-6153-434a-b411-194fc252c98c)

**Comment by dbannik** [comment]
### workaround
### version 2.16.1
cause flush as often as possible
**Before**
![image](https://github.com/doctrine/orm/assets/16108701/6ba6f664-f879-4e61-9eff-2c19d5c3633c)
**After**
![image](https://github.com/doctrine/orm/assets/16108701/2c55c0a8-ea02-487e-a553-f804574230db)

**Comment by dbannik** [comment]
i see a double insert
though there should only be one
![image](https://github.com/doctrine/orm/assets/16108701/d27fc311-33bc-45cb-8824-34eae17ffc26)

**Comment by dbannik** [comment]
The problem arises when **postPersist** event is caused and in it there is a flash

**Comment by jkgroupe** [comment]
Hello, I have the same issue but only when I'm trying to persist a `oneToMany` collection and only with `postPersist` (I don't have this issue with `postUpdate`).

I don't have this issue with my others listeners which using `flush` on `postPersist`

**My controller**
```
$receipt = new Receipt();
$receipt->setTitle('Test');

$period = new Period();
$period->setStartAt(new \DateTime);
$period->setEndAt(new \DateTime);

$receipt->addPeriod($period);

$this->entityManager->persist($receipt);
$this->entityManager->flush();
```


**My entity** 
```
/**
 * @ORM\Table(name="receipts")
 */
class Receipt implements ReceiptInterface
{
    /**
     * @var Collection<Period>
     * @ORM\OneToMany(targetEntity="App\Domain\Receipt\Period", mappedBy="receipt", cascade={"persist", "remove"}, fetch="EXTRA_LAZY", orphanRemoval=true)
     * @ORM\OrderBy({"orderNumber" = "ASC"})
     */
    protected Collection $periods;
}
```


**My Listener** 
```
#[AsEntityListener(event: Events::postPersist, method: 'postPersist', entity: Receipt::class)]
final class ReceiptListener 
{
     public function postPersist(Receipt $receipt, LifecycleEventArgs $args): void
     {
          if ($report = $receipt->getReport()) {
              $report->updateAmounts();
          }

          $args->getObjectManager()->flush();
     }
}
```

Hope it helps

**Comment by BrandonGillis** [comment]
Hello, I have the same issue when upgrading to `2.16.x`, as stated in the previous comment `flush()` shouldn't be used inside `postPersist` event, what would be the correct way to persist and flush another class entity that relies on the postPersist event to get the entity database-ID ?

**Comment by cbichis** [comment]
I confirm the same issue, on 2.16.x persisting of secondary entities doesnt seems to work. Which works on 2.15.x.

On 2.16.x no changes made on postPersist (other entities on which persist is called) doesnt seems to be flushed without the flush. And with the flush there is error regarding missing parameters.

Maybe this is related to?

https://github.com/doctrine/dbal/issues/5895

**Comment by codedge** [comment]
I run into this problem when using an [ApiPlatform state processor](https://api-platform.com/docs/core/state-processors/).
By persisting and flushing two entities, it generates one more INSERT statement with empty parameters.

```php
public function process($data, Operation $operation, array $uriVariables = [], array $context = []): Transaction
{
    $entityB = $this->entityBRepository->findOneBy(['refId' => $data->refId]);
    if ($entityB == null) {
        $entityB = (new EntityB())->setRefId($data->refId);
        $this->entityManager->persist($entityB);
    }

    $entity = new EntityA();
    $entity
        ->setEntityB($entityB)
        ->setReference($data->reference)
        ->setDate(new DateTimeImmutable($data->tDate))
    ;

    $this->entityManager->persist($entity);
    $this->entityManager->flush();

    return $entity;
}
```

**Comment by Seb33300** [comment]
Same issue here.

In the symfony profiler I can see the INSERT query properly executed the first time with all `?` replaced by parameters, and then, trying to execute the same query a second time without replacing `?`.

**Comment by Seb33300** [comment]
Reading previous comments, this may be related to listeners?
In our case we use https://github.com/doctrine-extensions/DoctrineExtensions on our entities (Timestampable)

**Comment by mpdude** [comment]
Everyone affected by this please try #10915 and report whether it helps

**Comment by yblatti** [comment]
Same problem for me.
In a symfony app, I use EntityListeners postPersist for multiple things : taking snapshots, updating ElasticSearch indices...
For all of this, I use symfony messenger, with doctrine transport : I dispatch a message on the bus in the postPersist(), now it fails every time.

> Everyone affected by this please try #10915 and report whether it helps

I tried it and it works for me !

**Comment by Herz3h** [comment]
I'm using gedmo tree and after persisting a tree, it is persisted in reverse. I had to revert to 2.15.5 for the bug to disappear. 2.17.2 didn't work for me.

**Comment by yblatti** [comment]
> I'm using gedmo tree and after persisting a tree, it is persisted in reverse. I had to revert to 2.15.5 for the bug to disappear. 2.17.2 didn't work for me.

Have a look at https://github.com/doctrine/orm/pull/11086, it's planned for [2.17.3](https://github.com/doctrine/orm/milestone/139).

**Comment by Herz3h** [comment]
Even upgrading to 2.18.1 doesn't fix the problem of reverse persist.

**Comment by mpdude** [comment]
@Herz3h can you provide a reproduce case?

**Comment by faton-joe** [comment]
Having this issue also with the latest stable version 2.19.4. The case is similar to this that might help to replicate.
Saving related entities.

```

       $category = new Category();
        $category->setName('Computer Peripherals');

        $product = new Product();
        $product->setName('Keyboard');
        $product->setPrice(19.99);
        $product->setDescription('Ergonomic and stylish!');

        // relates this product to the category
        $product->setCategory($category);

        $entityManager->persist($category);
        $entityManager->persist($product);
        $entityManager->flush();
```
        
 Now in case that I do flush() before setting the relation  `$product->setCategory($category);` this will go through, but I don't think this is the solution.
